### ISSUE TITLE
Any thoughts on making an "official" docker hub image?

### ISSUE THREAD

**Issue by truesilver92** [romanz/electrs#282]

**Comment by romanz** [comment]
Good question - as of today I prefer to have source releases, since I don't want to require more trust when downloading a binary/Docker release.
You can use the included [Dockerfile](https://github.com/romanz/electrs/blob/master/Dockerfile) to build your own image if needed.

**Comment by Kixunil** [comment]
Building `electrs` with `librocksdb` linked dynamically is deterministic, so we could both sign such binaries and anyone else interested could join. I already sign my deb packages but I understand most people will not want to manually unpack them (unless it becomes the preferred method of deployment), so I'm willing to sign binaries themselves if there's a demand.

**Comment by thomaseizinger** [comment]
> Good question - as of today I prefer to have source releases, since I don't want to require more trust when downloading a binary/Docker release. You can use the included [Dockerfile](https://github.com/romanz/electrs/blob/master/Dockerfile?rgh-link-date=2020-07-25T09%3A45%3A42Z) to build your own image if needed.

Would you be interested in a GitHub workflow that builds and attaches binaries to every release? I would like to deploy electrs to my raspberry and was thinking of creating a `-bin` AUR package to avoid compiling on it. It would be significantly easier if I could just download a binary from the release.

I understand that this requires trust in GitHub's CI service but maybe we can use it as a starting point? Tooling to verify that the build is in fact what we expect and as a result uploading a signature can be built on top of this.

**Comment by Kixunil** [comment]
@thomaseizinger if can get some people to reproduce and sign the releases then I'm in favor.

**Comment by thomaseizinger** [comment]
> @thomaseizinger if can get some people to reproduce and sign the releases then I'm in favor.

I am not sure I completely follow. Uploading a release-asset requires write access to the repository so I think it would have to be done by maintainers? Or are you thinking of a workflow where non-maintainers reproduce and sign and then request the signature to be uploaded to the release?

**Comment by Kixunil** [comment]
We could have some dedicated place to upload signatures for verified developers (to avoid spam).

**Comment by SRv6d** [comment]
In the spirit of  "[SO WHY THE FUCK IS THERE CODE? make an EXE file and give it to me. STUPID FUCKING SMELLY NERDS](https://github.com/sherlock-project/sherlock/issues/2011)", can this be reevaluated?
I would argue that having to compile code before you can use it is quite the hurdle for the average user. Can we make it easier to run your own node by offering compiled binaries?
As @thomaseizinger suggested, using GitHub actions and releases would be a great way to do this, given that it allows for signed build provenance and we implicitly already trust GitHub with the source code.

**Comment by Kixunil** [comment]
> we implicitly already trust GitHub with the source code.

We don't. @romanz signs the tags.

If you don't mind trusting GH you can extract binaries from CADR packages. CADR builds now have two builders - myself and GH actions. You can compare the hashes when you look at CADR jobs.

**Comment by SRv6d** [comment]
It's great that tags are signed, but when I git clone, GitHub **could** theoretically remove signatures and the only way for me to know that they were there in the first place would be (again) GitHub. Further, to verify the tags I **first** need to get @romanz PGP public key, which is in his GitHub profile or his website (again linked on GitHub).

Discussion about trust in GitHub aside, should I interpret your reply as "We are not currently interested in publishing official binaries or container images through GitHub"? Regarding CADAR, sorry but I've never heard that before. I'm fine with building myself, I was simply asking to make this more accessible to the general public.

**Comment by Kixunil** [comment]
> the only way for me to know that they were there in the first place would be (again) GitHub

You can ask people around. You can also open an issue and if you get no response you know the author is MIA or GH is doing something nefarious so you know to avoid the SW in either case. (I do this BTW.)

> Further, to verify the tags I **first** need to get @romanz PGP public key, which is in his GitHub profile or his website (again linked on GitHub).

You can go to a conference. :) But really, TOFU is much better than not verifying signatures at all so signature verification is still relevant.

> We are not currently interested in publishing official binaries or container images through GitHub

More like nobody had time to look into it properly yet.

> Regarding CADAR, sorry but I've never heard that before.

https://github.com/debian-cryptoanarchy/cryptoanarchy-deb-repo-builder/ (it's also linked in the docs).

**Comment by SRv6d** [comment]
Yes you can ask around or get public keys at conferences but I'm not sure how realistic of an expectation that is from the average user.
[CADAR](https://github.com/debian-cryptoanarchy/cryptoanarchy-deb-repo-builder/) seems like a good option, although still not really low friction in my opinion.

> More like nobody had time to look into it properly yet

If you are interested, I could contribute this in a PR, I've implemented attested binaries and containers built using GitHub workflows for plenty of my own repos. @thomaseizinger might also still be good for this.

**Comment by Kixunil** [comment]
> I'm not sure how realistic of an expectation that is from the average user.

Doesn't matter if TOFU is still better than not doing any verification at all.

> although still not really low friction in my opinion.

If you're using Debian already and trust me there's nothing with lower friction that I know of. If you just download the binary you don't get the automation CADR provides. You still have to verify signature for each update, configure `bitcoind` and electrs, debug if anything goes wrong... With CADR you just copy-paste a few commands and then use the normal installation software you'd use for any other app and you'll end up with the daemons running/managed automatically. The only thing that comes close to it is installing the `nix` packager and then installing nix-bitcoin there. But this has the additional step of installing a new packager which CADR doesn't have. (Of course, if you're already using `nix` then the situation is reversed!)

> If you are interested, I could contribute this in a PR, I've implemented attested binaries and containers built using GitHub workflows for plenty of my own repos

This is mainly a question for @romanz. I'm not sure how the attestation works but if it can be combined with deterministic builds it would be great. I'm willing to compile independently and publish my signature.
